File: src/contracts/event-topic.ts
export const EVENT_TOPICS = [
  "graph",
  "render",
  "provider",
  "optimizer",
  "worker",
  "question",
  "recovery",
  "budget",
  "audit",
  "runtime",
] as const;

export type EventTopic = (typeof EVENT_TOPICS)[number];

export const EVENT_TOPIC_ERRORS = ["UnknownTopic", "EmptyTopic"] as const;

export type EventTopicError = (typeof EVENT_TOPIC_ERRORS)[number];

export const PHASE_0A_LIVE_TOPICS = ["runtime"] as const satisfies readonly EventTopic[];

export class EventTopicParseError extends Error {
  readonly kind: EventTopicError;

  constructor(kind: EventTopicError) {
    super(kind);
    this.name = "EventTopicParseError";
    this.kind = kind;
  }
}

export function parseEventTopic(raw: string): EventTopic {
  if (raw === "") {
    throw new EventTopicParseError("EmptyTopic");
  }

  if (isEventTopic(raw)) {
    return raw;
  }

  throw new EventTopicParseError("UnknownTopic");
}

export function isPhase0aLiveTopic(topic: EventTopic): boolean {
  return (PHASE_0A_LIVE_TOPICS as readonly EventTopic[]).includes(topic);
}

function isEventTopic(value: string): value is EventTopic {
  return (EVENT_TOPICS as readonly string[]).includes(value);
}

File: src-tauri/src/contracts/event_topic.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum EventTopic {
    Graph,
    Render,
    Provider,
    Optimizer,
    Worker,
    Question,
    Recovery,
    Budget,
    Audit,
    Runtime,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum EventTopicError {
    UnknownTopic,
    EmptyTopic,
}

impl std::fmt::Display for EventTopicError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let value = serde_json::to_string(self).map_err(|_| std::fmt::Error)?;
        formatter.write_str(value.trim_matches('"'))
    }
}

impl std::error::Error for EventTopicError {}
